dkg/frost/frost.go

```diff
@@ -0,0 +1,756 @@
+			return nil, err
+		}
+
+		blameOutput := &dkg.BlameOutput{
+			Valid:        valid,
+			BlameMessage: blameMessageBytes,
+		}
+		return blameOutput, nil
+	}
+	return nil, nil
+}
+
+func (fr *FROST) processBlameTypeInvalidShare(operatorID uint32, blameMessage *BlameMessage) (bool /*valid*/, error) {
+	round1Message := &Round1Message{}
+	if err := json.Unmarshal(blameMessage.BlameData[0], round1Message); err != nil {
+		return false, err
+	}
+
+	verifiers := new(sharing.FeldmanVerifier)
+	for _, commitmentBytes := range round1Message.Commitment {
+		commitment, err := thisCurve.Point.FromAffineCompressed(commitmentBytes)
+		if err != nil {
+			return false, err
+		}
+		verifiers.Commitments = append(verifiers.Commitments, commitment)
+	}
+
+	blamerSessionSK := ecies.NewPrivateKeyFromBytes(blameMessage.BlamerSessionSk)
+	shareBytes, err := ecies.Decrypt(blamerSessionSK, round1Message.Shares[operatorID])
+	if err != nil {
+		return false, err
+	}
+
+	share := &sharing.ShamirShare{
+		Id:    operatorID,
+		Value: shareBytes,
+	}
+
+	if err := verifiers.Verify(share); err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func (fr *FROST) processBlameTypeInconsistentMessage(operatorID uint32, blameMessage *BlameMessage) (bool /*valid*/, error) {
+	if len(blameMessage.BlameData) != 2 {
+		return false, errors.New("invalid blame data")
+	}
+
+	var originalMessage, newMessage dkg.SignedMessage
+
+	if err := originalMessage.Decode(blameMessage.BlameData[0]); err != nil {
+		return false, err
+	}
+	if err := newMessage.Decode(blameMessage.BlameData[1]); err != nil {
+		return false, err
+	}
+
+	valid1 := originalMessage.Validate() == nil
+	valid2 := newMessage.Validate() == nil
+
+	return valid1 && valid2, nil
+}
+
+func (fr *FROST) canProceedRound1() bool {
+	if fr.currentRound != Preparation {
+		return false
+	}
+
+	for _, operatorID := range fr.operators {
+		protocolMessage := &ProtocolMsg{}
+
+		msg, ok := fr.msgs[Preparation][operatorID]
+		if ok {
+			if err := protocolMessage.Decode(msg.Message.Data); err != nil {
+				return false
+			}
+			if protocolMessage.PreparationMessage == nil {
+				return false
+			}
+		} else {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (fr *FROST) canProceedRound2() bool {
+	if fr.currentRound != Round1 {
+		return false
+	}
+
+	for _, operatorID := range fr.operators {
+		protocolMessage := &ProtocolMsg{}
+
+		msg, ok := fr.msgs[Round1][operatorID]
+		if ok {
+			if err := protocolMessage.Decode(msg.Message.Data); err != nil {
+				return false
+			}
+			if protocolMessage.Round1Message == nil {
+				return false
+			}
+		} else {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (fr *FROST) canProceedKeygenOutput() bool {
+	if fr.currentRound != Round2 {
+		return false
+	}
+
+	for _, operatorID := range fr.operators {
+		protocolMessage := &ProtocolMsg{}
+
+		msg, ok := fr.msgs[Round2][operatorID]
+		if ok {
+			if err := protocolMessage.Decode(msg.Message.Data); err != nil {
```

Comment:
why decoding? just checking all participants sent?

Reply:
yes, will remove this part as mentioned in previous comment